runtime/contract_update_test.go

```diff
@@ -0,0 +1,200 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package runtime
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/tests/utils"
+)
+
+func TestContractUpdateWithDependencies(t *testing.T) {
+
+	runtime := newTestInterpreterRuntime()
+	accountCodes := map[common.LocationID][]byte{}
+	signerAccount := common.MustBytesToAddress([]byte{0x1})
+
+	runtimeInterface := &testRuntimeInterface{
+		getCode: func(location Location) (bytes []byte, err error) {
+			return accountCodes[location.ID()], nil
+		},
+		storage: newTestLedger(nil, nil),
+		getSigningAccounts: func() ([]Address, error) {
+			return []Address{signerAccount}, nil
+		},
+		resolveLocation: singleIdentifierLocationResolver(t),
+		getAccountContractCode: func(address Address, name string) (code []byte, err error) {
+			location := common.AddressLocation{
+				Address: address,
+				Name:    name,
+			}
+			return accountCodes[location.ID()], nil
+		},
+		updateAccountContractCode: func(address Address, name string, code []byte) error {
+			location := common.AddressLocation{
+				Address: address,
+				Name:    name,
+			}
+			accountCodes[location.ID()] = code
+			return nil
+		},
+		emitEvent: func(event cadence.Event) error {
+			return nil
+		},
+		decodeArgument: func(b []byte, t cadence.Type) (value cadence.Value, err error) {
+			return json.Decode(b)
+		},
+
+		// Always force to get the old program from the source during the update.
+		getProgram: func(location Location) (*interpreter.Program, error) {
```

Comment:
👍 

Maybe we can also add a `setProgram` function to the interface, and then assert here and in `setProgram`, that during the update (after the initial deployment), the cache isn't read from or written to

Reply:
`setProgram` and `getProgram` do get called when parsing the 'current' program. So it may not be possible to isolate the cache read/writes during parsing the 'old-program' vs 'current-program' for assertion.
We could test 'what' is read from the cache, but I guess that's already tested by `TestRuntimeContractUpdateProgramCaching`?

Reply:
Right, they get called for the update. In the old version of the code, there were multiple calls to get/set program.
I tried to add a best-effort assertion for this in [02293b0](https://github.com/onflow/cadence/pull/1689/commits/02293b0fae79385fb0820318bd44581c8d8d93c0): assert get/set is only called once.

Reply:
looks good! Thanks for adding that